### PR TITLE
nixos/webhook: add support for a webhook service option

### PR DESCRIPTION
**Pull request by ghost**
###### Description of changes

This is a new PR for the abandoned #83132.
The PR is to address the comments in that PR and add automated testes.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by LoveIsGrief** [comment]
@roberth you reviewed the previous PR years ago. Would you like to review this one again? @W95Psp as the original author, if you have input, please do share!

File: nixos/modules/services/networking/webhook.nix [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        example = literalExample "pkgs.webhook";
+        description = mdDoc ''
+          Webhook package to use.
+        '';
+      };
+      user = mkOption {
+        type = types.str;
+        default = defaultUser;
+        description = mdDoc ''
+          Webhook will be run under this user (user will be created if it doesn't exist. This can be your user name).
```
Comment (roberth) [inline nixos/modules/services/networking/webhook.nix:36]:
```suggestion
          Webhook will be run under this user.
          
          This user will be created if the username is the default, `webhook`.
```

File: nixos/modules/services/networking/webhook.nix [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    )
+  );
+
+  # Todo: more expressive type?
+  typeTriggerRule = types.attrs;
+
+in {
+  options = {
+    services.webhook = {
+      enable = mkEnableOption (mdDoc "Webhook is server written in Go, that allows you to create HTTP endpoints (hooks), which execute configured commands for any person or service that knows the URL.");
```
Comment (roberth) [inline nixos/modules/services/networking/webhook.nix:21]: Please add a link.

File: nixos/modules/services/networking/webhook.nix [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        {
+          isSystemUser = true;
+          group = cfg.group;
+          description = "Webhook daemon user";
+        };
+    };
+
+    users.groups = mkIf (cfg.group == defaultUser) {
+      ${defaultUser} = {};
+    };
```
Comment (roberth) [inline nixos/modules/services/networking/webhook.nix:159]: What if only the user is set? In don't think we should create the group in that case.

File: nixos/modules/services/networking/webhook.nix [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+      group = mkOption {
+        type = types.str;
+        default = defaultUser;
+        description = mdDoc ''
+          Webhook will be run under this group (group will be created if it doesn't exist. This can be your group name).
+        '';
+      };
+      ip = mkOption {
+        type = types.str;
+        default = "0.0.0.0";
```
Comment (roberth) [inline nixos/modules/services/networking/webhook.nix:48]: This binds on all interfaces, but there's no `openFirewall` option.
Should `webhook` be locally reverse proxied or not?

File: nixos/modules/services/networking/webhook.nix [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        description = mdDoc "ip the webhook should serve hooks on";
+      };
+      port = mkOption {
+        type = types.port;
+        default = 9000;
+        description = mdDoc "port the webhook should serve hooks on";
+      };
+      template = mkOption {
```
Comment (roberth) [inline nixos/modules/services/networking/webhook.nix:56]:
```suggestion
      enableTemplate = mkOption {
```
+ change below

File: nixos/tests/services/webhook.nix [hunk 6]
```diff
@@ -0,0 +1,28 @@
+{ pkgs, ... }:
+{
+  name = "webhook";
+
+  nodes = {
+    webhookMachine = { pkgs, ... }: {
+      services.webhook = {
+        enable = true;
+      };
+    };
```
Comment (roberth) [inline nixos/tests/services/webhook.nix:10]: Maybe add a client node to test the `openFirewall` option?

File: nixos/modules/services/networking/webhook.nix [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+        type = types.nullOr types.ints.positive;
+        default = null;
+        description = mdDoc "specifies the HTTP status code to be returned upon success";
+      };
+
+
+      triggerRule = mkOption {
+        type = types.nullOr typeTriggerRule;
+        default = null;
+        description = mdDoc "specifies the list of headers that will be returned in HTTP response for the hook";
```
Comment (roberth) [inline nixos/modules/services/networking/webhook.nix:137]: Copy paste error?

Is it about https://github.com/adnanh/webhook/blob/master/docs/Hook-Rules.md ?

File: nixos/modules/services/networking/webhook.nix [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+  defaultUser = "webhook";
+
+  hooksFile = pkgs.writeText "webhook-config" (
+    builtins.toJSON (
+      builtins.attrValues cfg.hooks
+    )
+  );
+
+  # Todo: more expressive type?
+  typeTriggerRule = types.attrs;
```
Comment (roberth) [inline nixos/modules/services/networking/webhook.nix:16]:
```suggestion
  # https://github.com/adnanh/webhook/blob/master/docs/Hook-Rules.md
  # We might merge with "or" (perhaps?), but until then, allow only a single definition.
  typeTriggerRule = types.uniq types.attrs;
```

**Review by roberth — COMMENTED** [review]
Some suggestions in the file comments.
Also it would be nice to have more explanation and links in the option descriptions. They're usually full sentences.

**Comment by RaitoBezarius** [comment]
FWIW, there was https://github.com/NixOS/nixpkgs/pull/194980 which was semi-recent. :)

**Comment by LoveIsGrief** [comment]
@RaitoBezarius the author of that PR didn't respond, so I just assumed it was dead. I'll march forward with this one and if @pbsds  has feedback, it's very welcome. Some parts of his PR will be used as inspiration :slightly_smiling_face:

File: nixos/tests/services/webhook.nix (hunk 6 above)
Comment (LoveIsGrief) [reply nixos/tests/services/webhook.nix:10]: I was wondering why forwarding the port wasn't working... the manual doesn't mention how to communicate between nodes or if they're in the same network. I'll try and figure that out, but if it's not straight forward, a port-forward it'll be!

File: nixos/modules/services/networking/webhook.nix (hunk 4 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix:48]: Added an `openFirewall` option in https://github.com/NixOS/nixpkgs/pull/205636/commits/c5501723ef2ab062bccc7cdb8ecd433fc549434b

It'll be up to the user to decide whether they want to reverse proxy it or not

File: nixos/tests/services/webhook.nix (hunk 6 above)
Comment (LoveIsGrief) [reply nixos/tests/services/webhook.nix:18]: After searching for a while, I couldn't find documentation on communication between VM and went for the simple option of forwarding port a port on the host to the guest after adding the `openFirewall` option. To use a client machine, I'd need a code snippet to show how that's done.

Comment (roberth) [reply nixos/tests/services/webhook.nix:18]: Oh, what I was trying to say was to put another node in the `nodes` attribute; two VMs talking to each other.
That's how most networked tests are written.
I don't really know how port forwarding compares to having two VMs in a network.

File: nixos/modules/services/networking/webhook.nix (hunk 1 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix]: The descriptions should be better now

File: nixos/modules/services/networking/webhook.nix (hunk 3 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix]: OK, changes are in https://github.com/NixOS/nixpkgs/pull/205636/commits/9b34c109c687d27efac10ac7fe75549d4c7eafa1

File: nixos/modules/services/networking/webhook.nix (hunk 7 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix]: I removed it for now as it's not used anywhere atm. It could be introduced in another PR

File: nixos/modules/services/networking/webhook.nix [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+  # https://github.com/adnanh/webhook/blob/master/docs/Hook-Rules.md
+  # We might merge with "or" (perhaps?), but until then, allow only a single definition.
+  typeTriggerRule = types.uniq types.attrs;
+
+in {
+  options = {
+    services.webhook = {
+      enable = mkEnableOption (mdDoc ''
+        [Webhook] is server written in Go, that allows you to create HTTP endpoints (hooks),
```
Comment (RaitoBezarius) [inline nixos/modules/services/networking/webhook.nix:23]: Are you certain this is valid Markdown? I never knew about this syntax for links.

File: nixos/modules/services/networking/webhook.nix [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+        }));
+      };
+
+      successHttpResponseCode = mkOption {
+        type = types.nullOr types.ints.positive;
+        default = null;
+        description = mdDoc "This specifies the HTTP status code to be returned upon a successful hook call.";
+      };
+
+      # TODO: add other options from https://github.com/adnanh/webhook/blob/master/docs/Hook-Definition.md
```
Comment (RaitoBezarius) [inline nixos/modules/services/networking/webhook.nix:168]: Why not adopting a `freeformType`-style options to let people define everything they want without you needing to type everything?

File: nixos/modules/services/networking/webhook.nix [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+    };
+  };
+
+  imports = [];
```
Comment (RaitoBezarius) [inline nixos/modules/services/networking/webhook.nix:172]: This can be removed.

File: nixos/modules/services/networking/webhook.nix (hunk 9 above)
Comment (roberth) [reply nixos/modules/services/networking/webhook.nix:23]: Lgtm, but it's good to check. I think there's a formatting improvement to be made a bit further down ;)

**Comment by LoveIsGrief** [comment]
This is valid CommonMark called a shortcut reference link

https://spec.commonmark.org/0.30/#shortcut-reference-link

I don't know how to test whether https://search.nixos.org/options supports it, but if it uses CommonMark, it should. Is there documentation for that?

**Comment by LoveIsGrief** [comment]
This was how it was in the previous PR, but I agree, freeform might be better for posteriority as more options get added. Will modify.

File: nixos/modules/services/networking/webhook.nix (hunk 10 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix]: The hooks are now freeform

File: nixos/modules/services/networking/webhook.nix (hunk 9 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix:19]: This is valid CommonMark called a shortcut reference link

https://spec.commonmark.org/0.30/#shortcut-reference-link

I don't know how to test whether https://search.nixos.org/options supports it, but if it uses CommonMark, it should. Is there documentation for that?

File: nixos/modules/services/networking/webhook.nix (hunk 11 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix]: Done

File: nixos/modules/services/networking/webhook.nix (hunk 9 above)
Comment (roberth) [reply nixos/modules/services/networking/webhook.nix:19]: You could build the manual locally: https://nixos.org/manual/nixos/unstable/#chap-contributing
I don't know if search.nixos.org has extra demands. @ncfavier?

File: nixos/modules/services/networking/webhook.nix [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+        which execute configured commands for any person or service that knows the URL.
+
+        [Webhook]: https://github.com/adnanh/webhook
+      '');
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.webhook;
+        defaultText = "pkgs.webhook";
+        example = literalExample "pkgs.webhook";
```
Comment (ncfavier) [inline nixos/modules/services/networking/webhook.nix:29]:
```suggestion
        example = literalExpression "pkgs.webhook";
```

File: nixos/modules/services/networking/webhook.nix (hunk 9 above)
Comment (ncfavier) [reply nixos/modules/services/networking/webhook.nix:19]: It's fine, we use options.json which only has rendered docbook.

Comment (ncfavier) [reply nixos/modules/services/networking/webhook.nix:19]:
```suggestion
        [Webhook] is a server written in Go that allows you to create HTTP endpoints (hooks),
```

File: nixos/modules/services/networking/webhook.nix [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+          User = cfg.user;
+          Group = cfg.group;
+          ExecStart = ''
+            ${cfg.package}/bin/webhook \
+              -hooks ${hooksFile} \
+              ${if cfg.enableTemplate then "-template" else ""} \
+              -ip ${cfg.ip} \
+              -port ${toString cfg.port} \
+              -urlprefix ${cfg.urlPrefix}
+          # '';
```
Comment (pbsds) [inline nixos/modules/services/networking/webhook.nix:154]:
```suggestion
          '';
```

**Review by pbsds — APPROVED** [review]
LGTM, covers everything my PR did. Sorry for being unresponsive.

File: nixos/modules/services/networking/webhook.nix (hunk 12 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix:29]: Out of curiosity, what's the difference?

Comment (ncfavier) [reply nixos/modules/services/networking/webhook.nix:29]:
```
trace: warning: literalExample is deprecated, use literalExpression instead, or use literalDocBook for a non-Nix description.
```

File: nixos/modules/services/networking/webhook.nix (hunk 13 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix]: Done

File: nixos/modules/services/networking/webhook.nix (hunk 12 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix]: Thanks. Updated

File: nixos/modules/services/networking/webhook.nix (hunk 9 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix]: It works locally for me. I also applied @ncfavier's suggestion.

File: nixos/modules/services/networking/webhook.nix [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+
+let
+  cfg = config.services.webhook;
+  defaultUser = "webhook";
+
+  hooksFile = pkgs.writeText "webhook-config" cfg.hooks;
+
+  # https://github.com/adnanh/webhook/blob/master/docs/Hook-Rules.md
+  # We might merge with "or" (perhaps?), but until then, allow only a single definition.
+  typeTriggerRule = types.uniq types.attrs;
```
Comment (ncfavier) [inline nixos/modules/services/networking/webhook.nix:13]: This is not used

File: nixos/modules/services/networking/webhook.nix [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+        [Webhook] is a server written in Go that allows you to create HTTP endpoints (hooks),
+        which execute configured commands for any person or service that knows the URL.
+
+        [Webhook]: https://github.com/adnanh/webhook
+      '');
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.webhook;
+        defaultText = "pkgs.webhook";
```
Comment (ncfavier) [inline nixos/modules/services/networking/webhook.nix:28]:
```suggestion
        defaultText = literalExpression "pkgs.webhook";
```

File: nixos/modules/services/networking/webhook.nix [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+        description = mdDoc "Enable the generated hooks file to be parsed as a Go template.";
+      };
+      urlPrefix = mkOption {
+        type = types.str;
+        default = "hooks";
+        description = mdDoc ''
+          The URL path prefix to use for served hooks
+
+          Example: protocol://yourserver:port/PREFIX/:hook-id
+        '';
```
Comment (ncfavier) [inline nixos/modules/services/networking/webhook.nix:89]:
```suggestion
        example = "protocol://yourserver:port/PREFIX/:hook-id";
        description = mdDoc ''
          The URL path prefix to use for served hooks.
        '';
```

File: nixos/modules/services/networking/webhook.nix [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+        serviceConfig = {
+          Restart = "on-failure";
+          SuccessExitStatus = "2 3 4";
+          RestartForceExitStatus="3 4";
+          User = cfg.user;
+          Group = cfg.group;
+          ExecStart = ''
+            ${cfg.package}/bin/webhook \
+              -hooks ${hooksFile} \
+              ${if cfg.enableTemplate then "-template" else ""} \
```
Comment (ncfavier) [inline nixos/modules/services/networking/webhook.nix:151]:
```suggestion
              ${optionalString cfg.enableTemplate "-template"} \
```

File: nixos/modules/services/networking/webhook.nix [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+
+    systemd.services = {
+      webhook = {
+        description = "Webhook service";
+        after = [ "network.target" ];
+        environment = {
+          STNORESTART = "yes";
+          STNOUPGRADE = "yes";
+          inherit (cfg);
+        } // config.networking.proxy.envVars;
```
Comment (ncfavier) [inline nixos/modules/services/networking/webhook.nix:140]: This looks like it was cargo-culted from the Syncthing service.

```suggestion
        environment = config.networking.proxy.envVars;
```

File: nixos/modules/services/networking/webhook.nix [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+        wantedBy = [ "multi-user.target" ];
+        serviceConfig = {
+          Restart = "on-failure";
+          SuccessExitStatus = "2 3 4";
+          RestartForceExitStatus="3 4";
```
Comment (ncfavier) [inline nixos/modules/services/networking/webhook.nix:145]: Is this needed or was it also cargo-culted from the Syncthing service? What do these exit statuses mean for webhook?

File: nixos/modules/services/networking/webhook.nix [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+
+in {
+  options = {
+    services.webhook = {
+      enable = mkEnableOption (mdDoc ''
+        [Webhook] is a server written in Go that allows you to create HTTP endpoints (hooks),
+        which execute configured commands for any person or service that knows the URL.
+
+        [Webhook]: https://github.com/adnanh/webhook
+      '');
```
Comment (ncfavier) [inline nixos/modules/services/networking/webhook.nix:23]: Since this is a `mkEnableOption` the text will be wrapped in `Whether to enable {}.`. You could either adapt it to fit that format, override the `description`, or just use `mkOption`.

**Review by ncfavier — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/webhook.nix [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+          User = cfg.user;
+          Group = cfg.group;
+          ExecStart = ''
+            ${cfg.package}/bin/webhook \
+              -hooks ${hooksFile} \
+              ${if cfg.enableTemplate then "-template" else ""} \
+              -ip ${cfg.ip} \
+              -port ${toString cfg.port} \
+              -urlprefix ${cfg.urlPrefix}
+          '';
```
Comment (ncfavier) [inline nixos/modules/services/networking/webhook.nix:155]: We might want to add an `extraArgs` option for arguments that are not supported by the module.

File: nixos/modules/services/networking/webhook.nix (hunk 18 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix:139]: Probably was. Applying suggestion

File: nixos/modules/services/networking/webhook.nix (hunk 19 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix:140]: Syncthing probably served as a template in the original PR.

From a cursory glance at webhook, it seems like `0` is the success exit status and `1` is the only other one that I could find: https://github.com/adnanh/webhook/search?q=exit

image

I'll use those

File: nixos/modules/services/networking/webhook.nix (hunk 20 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix:23]: I'll adapt the text to fix the `Whether to enable` prefix.

File: nixos/modules/services/networking/webhook.nix (hunk 19 above)
Comment (ncfavier) [reply nixos/modules/services/networking/webhook.nix:140]: We should just remove the three options. Those are the default exit status semantics, and there's no reason to restart on failure a priori.

File: nixos/modules/services/networking/webhook.nix (hunk 21 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix:162]: Done in  1fa6933

File: nixos/modules/services/networking/webhook.nix (hunk 20 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix]: Done in c7c7371

File: nixos/modules/services/networking/webhook.nix (hunk 19 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix]: Applied in 2a9220c

File: nixos/modules/services/networking/webhook.nix (hunk 14 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix]: Removed in 3d19e72

File: nixos/modules/services/networking/webhook.nix (hunk 19 above)
Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix]: Oh, I didn't see your response. In any case, I do think that it should be restarted on failure. It's akin to a web-server and IMO those shouldn't just die forever when things go wrong.

If you like, I could introduce an option `restartOnFailure` or something similar to give users the power to decide for themselves. Or if you're fine with things as they are, you can leave this resolved.

File: nixos/modules/services/networking/webhook.nix [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+      urlPrefix = mkOption {
+        type = types.str;
+        default = "hooks";
+        example = "protocol://yourserver:port/PREFIX/:hook-id";
+        description = mdDoc ''
+          The URL path prefix to use for served hooks.
+        '';
+      };
+      hooks = mkOption {
+        type = types.str;
```
Comment (ncfavier) [inline nixos/modules/services/networking/webhook.nix:87]: I'm thinking this option's type should probably be JSON instead of an opaque string, or maybe something even more specific like

```nix
let hooksFormat = pkgs.formats.json {}; in
types.attrsOf (types.submodule ({ name, ... }: {
  freeformType = hooksFormat.type;
  options = {
    id = {
      type = types.str;
      default = name;
    };
    execute-command = {
      type = types.str;
    };
  };
}))
```

...and then use `hooksFormat.generate "webhook-config.json" (builtins.attrValues cfg.hooks)` or something like that. This makes it easier to define hooks across multiple modules, and we can also add an assertion checking that at least one hook is defined.

See https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md for more information.

I think this should be done in this PR so that we get the interface right the first time and don't have to do painful migrations in the future. Don't hesitate to ask if you need help implementing this.

One thing to ponder is how this interacts with [`-template`](https://github.com/adnanh/webhook/blob/master/docs/Templates.md). It seems like in typical usage, the template syntax would be confined inside JSON strings, so there should no problem translating it to Nix syntax. If more flexibility turns out to be needed, we can always add a way to configure a hook using a single string in the future.

File: nixos/modules/services/networking/webhook.nix [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+        type = types.package;
+        default = pkgs.webhook;
+        defaultText = literalExpression "pkgs.webhook";
+        example = literalExpression "pkgs.webhook";
+        description = mdDoc ''
+          Which Webhook package to use
+
+          This allows passing different version of the package.
+        '';
+      };
```
Comment (ncfavier) [inline nixos/modules/services/networking/webhook.nix:31]:
```suggestion
      package = mkPackageOption pkgs "webhook" {};
```

Sorry to come back to this a third time 😅

Comment (ncfavier) [inline nixos/modules/services/networking/webhook.nix:31]:
```suggestion
      package = mkPackageOption pkgs "webhook" {};
```

Sorry to come back to this a third time 😅

File: nixos/modules/services/networking/webhook.nix (hunk 22 above)
Comment (ncfavier) [reply nixos/modules/services/networking/webhook.nix:87]: > the template syntax would be confined inside JSON strings, so there should no problem translating it to Nix syntax

I guess there are problems with escaping though, e.g.
```
"secret": "{{ getenv "XXXTEST_SECRET" | js }}",
```
isn't valid JSON, so we can't generate that with `pkgs.formats.json`.

It seems like webhook supports passing multiple `-hooks` flags, so we could have one for each hook, and each hook could be either a structured submodule or a literal string.

```nix
hooks.hello = { execute-command = "hello"; };
hooks.helloTemplate = ''{"id": "hello-template", "execute-command": "{{hello "template" | whatever}}"}'';
```

Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix:75]: This was modified in favor of a string in 3c65372ac93a61ba1a38c9920869f3d92b6aedf2 as per [this comment](https://github.com/NixOS/nixpkgs/pull/205636#discussion_r1051495119) in the current PR.

Thank you very much for the comments. I learned more about options and discovered that probably a few things could be automated e.g checking that `enable` and `package` are of certain option types. 
To be honest though, I'm kind of mentally done with this PR and am going to move on to other things I was planning on working on and learning during the holiday season. Reading nixos RFCs weren't on my list :wink:

Comment (ncfavier) [reply nixos/modules/services/networking/webhook.nix:75]: Oh, then it seems like we can start with reverting that commit and adding a `freeformType` as suggested.

> To be honest though, I'm kind of mentally done with this PR

That's fine, I can take over. Thanks for the work done!

Comment (LoveIsGrief) [reply nixos/modules/services/networking/webhook.nix:75]: > That's fine, I can take over. Thanks for the work done!

image

You have edit access on the branch. Thanks for taking over

**Review by ncfavier — APPROVED** [review]
I've added a separate `hooksTemplated` option for literal hooks. Setting this option implies `enableTemplate`.

I did not use `either hookType str` because of https://github.com/NixOS/nixpkgs/issues/207962.

I think this is ready now, though a final review wouldn't hurt.

**Comment by ncfavier** [comment]
@ofborg test webhook

File: nixos/doc/manual/release-notes/rl-2305.section.md [hunk 24]
```diff
@@ -34,7 +36,7 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - `carnix` and `cratesIO` has been removed due to being unmaintained, use alternatives such as [naersk](https://github.com/nix-community/naersk) and [crate2nix](https://github.com/kolloch/crate2nix) instead.
 
-- `borgbackup` module now has an option for inhibiting system sleep while backups are running, defaulting to off (not inhibiting sleep), available as [`services.borgbackup.jobs.<name>.inhibitsSleep`](#opt-services.borgbackup.jobs._name_.inhibitsSleep). 
+- `borgbackup` module now has an option for inhibiting system sleep while backups are running, defaulting to off (not inhibiting sleep), available as [`services.borgbackup.jobs.<name>.inhibitsSleep`](#opt-services.borgbackup.jobs._name_.inhibitsSleep).
```
Comment (RaitoBezarius) [inline nixos/doc/manual/release-notes/rl-2305.section.md:39]: please undo this unrelated change

**Review by RaitoBezarius — APPROVED** [review]
LGTM.

File: nixos/doc/manual/release-notes/rl-2305.section.md (hunk 24 above)
Comment (ncfavier) [reply nixos/doc/manual/release-notes/rl-2305.section.md:39]: That reminds me i should set up editorconfig. Apparently we keep trailing whitespace for markdown because you can use it to insert line breaks? Fascinating

**Comment by RaitoBezarius** [comment]
@ofborg test webhook

**Comment by ncfavier** [comment]
Backport: https://github.com/NixOS/nixpkgs/pull/208547
